File: src/main/java/seremis/geninfusion/soul/traits/TraitTexture.java
package seremis.geninfusion.soul.traits;

import net.minecraft.util.DamageSource;
import net.minecraft.util.ResourceLocation;
import seremis.geninfusion.api.soul.IEntitySoulCustom;
import seremis.geninfusion.api.soul.SoulHelper;
import seremis.geninfusion.api.soul.lib.Genes;
import seremis.geninfusion.helper.GITextureHelper;

public class TraitTexture extends Trait {

    @Override
    public String getEntityTexture(IEntitySoulCustom entity) {
        try {
            return SoulHelper.geneRegistry().getValueString(entity, Genes.GENE_TEXTURE);
        } catch(NullPointerException e) {
            return "textures/entity/zombie/zombie.png";
        }
    }

    @Override
    public void onDeath(IEntitySoulCustom entity, DamageSource source) {
        //TODO if it doesn't spawn a Soul, remove the texture
        GITextureHelper.deleteTexture(toResource(getEntityTexture(entity)));
        GITextureHelper.deleteTexture(toResource((String) SoulHelper.geneRegistry().getChromosomeFor(entity, Genes.GENE_TEXTURE).getRecessive().getAlleleData()));
    }

    public ResourceLocation toResource(String string) {
        return new ResourceLocation(string);
    }
}

File: src/main/java/seremis/geninfusion/soul/traits/TraitTexture.scala
package seremis.geninfusion.soul.traits

import net.minecraft.util.{DamageSource, ResourceLocation}
import seremis.geninfusion.api.soul.lib.Genes
import seremis.geninfusion.api.soul.{IEntitySoulCustom, SoulHelper}
import seremis.geninfusion.helper.GITextureHelper
import seremis.geninfusion.lib.Localizations

class TraitTexture extends Trait {

    override def getEntityTexture(entity: IEntitySoulCustom): String = {
        try {
            SoulHelper.geneRegistry.getValueFromAllele[String](entity, Genes.GENE_TEXTURE)
        } catch {
            case e: NullPointerException => Localizations.LocModelTextures + Localizations.Blank
        }
    }

    override def onDeath(entity: IEntitySoulCustom, source: DamageSource) {
        //TODO make some texture thingy to ensure no infinite textures
        if(!entity.getWorld.isRemote && !entity.getSoulPreserved) {
            GITextureHelper.deleteTexture(toResource(getEntityTexture(entity)))
            GITextureHelper.deleteTexture(toResource(SoulHelper.geneRegistry.getChromosomeFor(entity, Genes.GENE_TEXTURE).get.getRecessive.getAlleleData.asInstanceOf[String]))
        }
    }

    def toResource(string: String): ResourceLocation = new ResourceLocation(string)
}